middleware/common/source/quantity.cpp

```diff
@@ -0,0 +1,120 @@
+//! 
+//! Copyright (c) 2024, The casual project
+//!
+//! This software is licensed under the MIT license, https://opensource.org/licenses/MIT
+//!
+#include "common/quantity.h"
+
+#include "common/algorithm.h"
+#include "common/string.h"
+#include "common/range.h"
+
+#include "common/string/compose.h"
+
+#include "common/code/raise.h"
+#include "common/code/casual.h"
+
+#include "casual/assert.h"
+
+#include <string>
+#include <map>
+#include <optional>
+
+namespace casual
+{
+   namespace common::quantity
+   {
+
+      namespace bytes
+      {
+         namespace local
+         {
+            namespace
+            {
+               namespace parse
+               {
+                  std::optional< bytes_type> symbol( std::string_view symbol)
+                  {
+                     static const std::map< std::string_view, bytes_type> symbol_map{
+                        { "", unit::byte },
+                        { symbol::byte, unit::byte },
+                        { symbol::kilobyte, unit::kilobyte },
+                        { symbol::kibibyte, unit::kibibyte },
+                        { symbol::megabyte, unit::megabyte },
+                        { symbol::mebibyte, unit::mebibyte },
+                        { symbol::gigabyte, unit::gigabyte },
+                        { symbol::gibibyte, unit::gibibyte },
+                        { symbol::terabyte, unit::terabyte },
+                        { symbol::tebibyte, unit::tebibyte }
+                     };
+
+                     if( auto found = symbol_map.find( symbol); found != std::end( symbol_map))
+                        return found->second;
+
+                     return std::nullopt;
+                  }
+               } // parse
+
+               namespace string
+               {
+                  std::string representation( bytes_type value)
+                  {
+                     static const std::map< bytes_type, std::string_view> unit_map{
+                        { unit::byte, symbol::byte},
+                        { unit::kilobyte, symbol::kilobyte},
+                        { unit::kibibyte, symbol::kibibyte},
+                        { unit::megabyte, symbol::megabyte},
+                        { unit::mebibyte, symbol::mebibyte},
+                        { unit::gigabyte, symbol::gigabyte},
+                        { unit::gibibyte, symbol::gibibyte},
+                        { unit::terabyte, symbol::terabyte},
+                        { unit::tebibyte, symbol::tebibyte}
+                     };
```

Comment:
You could just have a _mapping type_ and use a constexpr `std::array`, and just linearly `algorithm::find`.
```cpp
struct map
{
   bytes_type unit;
   std::string_view symbol;
   
   inline bool operator == ( const map& lhs, bytes_type rhs) { return lhs.unit == rhs;)
   inline bool operator == ( const map& lhs, std::string_view rhs) { return lhs.symbol == rhs;)
};
```
And use the array for both to and from string.